File: elixir/custom-set/custom_set.exs
defmodule CustomSet do
  defstruct list: []

  def delete(set, value), do: new(Enum.reject(set.list, &(&1 === value)))

  def difference(s1, s2), do: new(s1.list -- s2.list)

  def disjoint?(s1, s2), do: size(intersection(s1, s2)) == 0

  def empty(_set), do: new()

  def equal?(s1, s2), do: s1.list == s2.list

  def intersection(s1, s2), do: new(Enum.filter(s1.list, &(member?(s2, &1))))

  def member?(set, value), do: Enum.any?(set.list, &(&1 === value))

  def new(), do: %CustomSet{}
  def new([]), do: %CustomSet{}
  def new(set), do: %CustomSet{list: set |> Enum.uniq |> Enum.sort}

  def put(%CustomSet{list: values}, value), do: new([value|values])

  def size(set), do: Enum.count(set.list)

  def subset?(s1, s2), do: Enum.all?(s1.list, &(member?(s2, &1)))

  def to_list(set), do: set.list

  def union(s1, s2), do: new(s1.list ++ s2.list)
end

File: ruby/custom-set/custom_set.rb
class CustomSet
  attr_reader :set

  def initialize(set)
    @set = set
  end

  def empty?
    @set.empty?
  end

  def member?(item)
    @set.include?(item)
  end

  def subset?(other)
    @set.all? { |i| other.set.any? { |j| j.eql?(i) } }
  end

  def disjoint?(other)
    intersection(other).empty?
  end

  def intersection(other)
    CustomSet.new(@set.select { |i| other.member?(i) })
  end

  def difference(other)
    CustomSet.new(@set.select { |i| !other.member?(i) })
  end

  def union(other)
    out = CustomSet.new(@set)
    other.set.each { |i| out.add(i) }
    out
  end

  def ==(other)
    size == other.size && subset?(other)
  end

  def size
    @set.length
  end

  def add(item)
    @set.push(item).uniq!
    self
  end
end

module BookKeeping
  VERSION = 1
end
